### PR TITLE
ReactApplication should not impose a ReactRootView

### LINKED ISSUE #4925: ReactApplication doesn't have a way to mix RN and non-RN content
The playground app does a good job of showing how you can use `ReactNativeHost` .
The e2e test app uses `ReactApplication` - but it doesn't allow you to mix and match content between XAML or anything else, and RN.

The `Frame.Content` gets set to a new `ReactRootView` in the `ReactApplication`. This class has logic to create a Frame and a `ReactRootView` and everything works fine, but if I reparent that `ReactRootView` underneath e.g. a StackPanel, weird things start happening:

```cs
 base.OnLaunched(e);
 var frame = Window.Current.Content as Frame;
 if (useRNTester)
 {
     // inject the RNTester tree dump control
     var stackPanel = new StackPanel();
     var treeDumpControl = new TextBlock() { Text = "TreeDump", Padding = new Thickness(20) };
     stackPanel.Children.Add(treeDumpControl);
     stackPanel.Children.Add(frame.Content as ReactRootView);
     frame.Content = stackPanel;
}
```

Before the bundle loads everything looks ok - the textbox shows on top, the rest of the app underneath it.
image
But when the bundle loads, layout goes awry:
image

and if I create my own ReactRootView and add that to my layout, the React content isnt shown
What's the "right way" to mix RN and non-RN content?

From @vmoroz:

> The current ReactRootView is my very rough translation of what we had before in the ReactUWP. The only improvement that I had done for it is to make sure that we do not use shared_ptr for the C++/WinRT controls. Thus, the split of functionality between the ReactRootView and ReactRootControl. It was an anti-pattern and it had to be fixed. All shared_ptr ownership is in the ReactRootControl, while ReactRootView is a true XAML C++/WinRT control. I guess if something does not work properly, then it could be related to how these two communicate between each other. 
Also, in Office version we unload user UI when UI control get un-parented. I do not remember if I implemented this mechanism in RNW (I will verify it later today). It means that the UI must be reloaded after it is parented again. Though, I would expect that it must be done automatically.

Also it has a quite strange hierarchy:
```txt
// Visual tree to support safe harbor
  // m_rootView
  //  safe harbor
  //  m_xamlRootView
  //    JS created children
```
> The ReactApplication is used in the new applications. It internally initializes ReactRootView. The ReactRootView can be used on its own along with the ReactNativeHost if developers want to embed the RN into their existing apps. The ReactApplication is considered as a "utility" class for these two and to get a simple RN app quickly up-and-running.
No, I do not see that the unloading of RN UI on component un-parenting be implemented in RNW.
The only other explanation for the behavior that you see is that we may not have correct attributes for resizing of the ReactRootView vs other components you have. Is it possible that it becomes zero height? When I worked with it and saw some UI rendering issues, I had to add a border thickness and color to see what is happening.

### LINKED ISSUE #4940: Make app template use ReactNativeHost
When we discussed #4925 we mentioned that having the template use `ReactApplication` adds friction when a developer realizes that they want to add some XAML content to their app since `ReactApplication` doesn't allow this (it doesn't preclude it, you don't get a crash/exception, it just doesn't work).

This issue tracks replacing the use of `ReactApplication` in the app project template so that it uses `ReactNativeHost` instead and it's easier to move between pure RN and mix of RN and XAML

### PR DESCRIPTION
**Pull request by asklar**
Fixes #4925 and allows mixing RN content with non-RN in a markup-first way, fixes #4940 
Fixes the exception I left in for #4619 now that the floating point bug is fixed.

image

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/react-native-windows/pull/4949)

### CODE REVIEW

File: packages/E2ETest/windows/ReactUWPTestApp/MainPage.xaml.cs [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    /// An empty page that can be used on its own or navigated to within a Frame.
+    /// </summary>
+    public sealed partial class MainPage : Page
+    {
+        public MainPage()
+        {
+            this.InitializeComponent();
+            var app = Application.Current as App;
+            myRootView.ReactNativeHost = app.Host;
+            app.Host.ReloadInstance();
```
Comment (vmoroz) [inline packages/E2ETest/windows/ReactUWPTestApp/MainPage.xaml.cs:31]: >app.Host.ReloadInstance(); [](start = 12, length = 26)

It should not be here. We want to load instance before the window is created. Reloading instance on UI creation will force us to make UI loading slower.

File: vnext/Microsoft.ReactNative/ReactApplication.h [hunk 2]
```diff
@@ -81,15 +81,11 @@ struct ReactApplication : NoDefaultCtorReactApplication_base<ReactApplication> {
...
 
  protected:
-  virtual ReactApplicationDelegate __stdcall CreateReactApplicationDelegate();
-
  private:
   ReactNative::ReactInstanceSettings m_instanceSettings{nullptr};
   Windows::Foundation::Collections::IVector<IReactPackageProvider> m_packageProviders{nullptr};
   ReactNative::ReactNativeHost m_host{nullptr};
 
-  ReactApplicationDelegate m_delegate{nullptr};
```
Comment (vmoroz) [inline vnext/Microsoft.ReactNative/ReactApplication.h:91]: > ReactApplicationDelegate m_delegate{nullptr}; [](start = 1, length = 46)

I think it is a good direction that we start to remove the ReactApplicationDelegate. It is better to have a simple ReactApplication and extend it. As a result the ReactApplicationDelegate code should be completely removed and good parts of it moved to ReactApplication.

File: vnext/Microsoft.ReactNative/ReactApplication.cpp [hunk 3]
```diff
@@ -131,16 +132,19 @@ void ReactApplication::OnLaunched(LaunchActivatedEventArgs const &e) {
...
 /// </summary>
 /// <param name="e">Details about the launch request and process.</param>
 void ReactApplication::OnCreate(IActivatedEventArgs const &e) {
-  if (!m_delegate) {
-    m_delegate = CreateReactApplicationDelegate();
-  }
+  // Resuming({this, &ReactApplicationDelegate::OnResuming});
+  // Suspending({this, &ReactApplicationDelegate::OnSuspending});
+  // LeavingBackground({this, &ReactApplicationDelegate::OnLeavingBackground});
+  // EnteredBackground({this, &ReactApplicationDelegate::OnEnteredBackground});
```
Comment (vmoroz) [inline vnext/Microsoft.ReactNative/ReactApplication.cpp:147]: It would be great if we can redirect all these events to the ReactNotificationService for which I am going to send a PR today.

File: vnext/Microsoft.ReactNative/ReactApplication.cpp [hunk 4]
```diff
@@ -179,34 +183,13 @@ void ReactApplication::OnCreate(IActivatedEventArgs const &e) {
...
-      // Ensure the current window is active
-      Window::Current().Activate();
-    }
   }
+  //
+  //
+  ApplyArguments(Host(), args.c_str());
+
+  // Nudge the ReactNativeHost to create the instance and wrapping context
+  Host().ReloadInstance();
```
Comment (vmoroz) [inline vnext/Microsoft.ReactNative/ReactApplication.cpp:192]: >Host().ReloadInstance(); [](start = 2, length = 24)

This seems to be the right place. Though, we can it event earlier if we could. Ideally it should be done as soon as we have all settings ready.

File: vnext/Microsoft.ReactNative/ReactApplication.cpp (hunk 3 above)
Comment (asklar) [reply vnext/Microsoft.ReactNative/ReactApplication.cpp:164]: @vmoroz can you clarify what you'd suggest? the methods in the delegate weren't really doing anything

Comment (vmoroz) [reply vnext/Microsoft.ReactNative/ReactApplication.cpp:164]: In future we can raise these events from the ReactNotificationSevice. For now, I would propose to delete this code to avoid having commented out "dead" code.

---
In reply to: [429664370](https://github.com/microsoft/react-native-windows/pull/4949#discussion_r429664370) [](ancestors = 429664370)

File: packages/E2ETest/windows/ReactUWPTestApp/MainPage.xaml.cs [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+using System.Runtime.InteropServices.WindowsRuntime;
+using Windows.Foundation;
+using Windows.Foundation.Collections;
+using Windows.UI.Xaml;
+using Windows.UI.Xaml.Controls;
+using Windows.UI.Xaml.Controls.Primitives;
+using Windows.UI.Xaml.Data;
+using Windows.UI.Xaml.Input;
+using Windows.UI.Xaml.Media;
+using Windows.UI.Xaml.Navigation;
```
Comment (vmoroz) [inline packages/E2ETest/windows/ReactUWPTestApp/MainPage.xaml.cs:15]: >using [](start = 0, length = 5)

Please use VS IDE to remove unused namespaces.

File: packages/microsoft-reactnative-sampleapps/windows/SampleAppCPP/packages.config [hunk 6]
```diff
@@ -1,5 +1,9 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <packages>
+  <package id="Microsoft.Net.UWPCoreRuntimeSdk" version="2.2.10" targetFramework="native" />
   <package id="Microsoft.UI.Xaml" version="2.3.191129002" targetFramework="native" />
   <package id="Microsoft.Windows.CppWinRT" version="2.0.200316.3" targetFramework="native" />
-</packages>
+  <package id="runtime.win10-arm.Microsoft.Net.UWPCoreRuntimeSdk" version="2.2.10" targetFramework="native" />
```
Comment (vmoroz) [inline packages/microsoft-reactnative-sampleapps/windows/SampleAppCPP/packages.config:6]: >package [](start = 3, length = 7)

It seems to be the missing part in my PR: [https://github.com/microsoft/react-native-windows/pull/4982](https://github.com/microsoft/react-native-windows/pull/4982).

File: packages/microsoft-reactnative-sampleapps/windows/SampleAppCPP/SampleAppCpp.vcxproj [hunk 7]
```diff
@@ -1,6 +1,10 @@
 <?xml version="1.0" encoding="utf-8"?>
 <!-- Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT License. See LICENSE in the project root for license information. -->
 <Project ToolsVersion="16.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <Import Project="..\packages\runtime.win10-arm.Microsoft.Net.UWPCoreRuntimeSdk.2.2.10\build\runtime.win10-arm.Microsoft.Net.UWPCoreRuntimeSdk.props" Condition="Exists('..\packages\runtime.win10-arm.Microsoft.Net.UWPCoreRuntimeSdk.2.2.10\build\runtime.win10-arm.Microsoft.Net.UWPCoreRuntimeSdk.props')" />
```
Comment (vmoroz) [inline packages/microsoft-reactnative-sampleapps/windows/SampleAppCPP/SampleAppCpp.vcxproj:4]: >Import [](start = 3, length = 6)

Is it correct to put these props files here?
Normally in C++ project they must go below.

File: packages/microsoft-reactnative-sampleapps/windows/SampleAppCPP/SampleAppCpp.vcxproj [hunk 8]
```diff
@@ -19,11 +23,11 @@
     <!-- Start Custom .NET Native properties -->
     <DotNetNativeVersion>2.2.7-rel-27913-00</DotNetNativeVersion>
     <!-- The name 'DotNetNativeVersion' is critical for restoring the right .NET framework libraries -->
-    <UWPCoreRuntimeSdkVersion>2.2.9</UWPCoreRuntimeSdkVersion>
+    <UWPCoreRuntimeSdkVersion>2.2.10</UWPCoreRuntimeSdkVersion>
   </PropertyGroup>
   <!-- Start Custom .NET Native targets -->
   <!-- Import all of the .NET Native / CoreCLR props at the beginning of the project -->
-  <Import Condition="'$(WindowsTargetPlatformMinVersion)' &gt;= '10.0.16299.0'" Project="$(ProgramFiles)\Microsoft SDKs\UWPNuGetPackages\Microsoft.Net.UWPCoreRuntimeSdk\$(UWPCoreRuntimeSdkVersion)\build\Microsoft.Net.UWPCoreRuntimeSdk.props" />
+  <!-- <Import Condition="'$(WindowsTargetPlatformMinVersion)' &gt;= '10.0.16299.0'" Project="$(ProgramFiles)\Microsoft SDKs\UWPNuGetPackages\Microsoft.Net.UWPCoreRuntimeSdk\$(UWPCoreRuntimeSdkVersion)\build\Microsoft.Net.UWPCoreRuntimeSdk.props" />
```
Comment (vmoroz) [inline packages/microsoft-reactnative-sampleapps/windows/SampleAppCPP/SampleAppCpp.vcxproj:30]: > <!-- [](start = 1, length = 5)

Could we rather have changes in this file be part of another PR? They seem here to be out-of-place, and not complete.
Plus, there is some conflict with my PR that is in progress.

File: vnext/local-cli/generator-windows/templates/cpp/src/App.cpp [hunk 9]
```diff
@@ -38,3 +42,89 @@ App::App() noexcept
...
 }
+
+/// <summary>
+/// Invoked when the application is launched normally by the end user.  Other entry points
+/// will be used such as when the application is launched to open a specific file.
+/// </summary>
+/// <param name="e">Details about the launch request and process.</param>
+void App::OnLaunched(LaunchActivatedEventArgs const& e)
+{
+    __super::OnLaunched(e);
```
Comment (vmoroz) [inline vnext/local-cli/generator-windows/templates/cpp/src/App.cpp:53]: >__super [](start = 4, length = 7)

This is not standard C++. Please do not use. 
Typically we just define an alias in header file:
using Super = ...
and the use it in code.

File: vnext/local-cli/generator-windows/templates/cpp/src/App.cpp [hunk 10]
```diff
@@ -38,3 +42,89 @@ App::App() noexcept
...
+            }
+            // Place the frame in the current Window
+            Window::Current().Content(rootFrame);
+            // Ensure the current window is active
+            Window::Current().Activate();
+        }
+    }
+    else
+    {
+        if (e.PrelaunchActivated() == false)
```
Comment (vmoroz) [inline vnext/local-cli/generator-windows/templates/cpp/src/App.cpp:95]: > == false [](start = 34, length = 9)

I wonder if the default XAML template has code like this. It is not a C++ style...

File: vnext/local-cli/generator-windows/templates/cpp/src/App.cpp [hunk 11]
```diff
@@ -38,3 +42,89 @@ App::App() noexcept
 
     InitializeComponent();
 }
+
+/// <summary>
+/// Invoked when the application is launched normally by the end user.  Other entry points
+/// will be used such as when the application is launched to open a specific file.
+/// </summary>
+/// <param name="e">Details about the launch request and process.</param>
+void App::OnLaunched(LaunchActivatedEventArgs const& e)
```
Comment (vmoroz) [inline vnext/local-cli/generator-windows/templates/cpp/src/App.cpp:51]: >OnLaunched [](start = 10, length = 10)

I wonder why the C++ OnLaunch code is so much more complex than C#.

File: vnext/Microsoft.ReactNative/ReactApplication.cpp [hunk 12]
```diff
@@ -85,14 +86,6 @@ ReactNative::ReactNativeHost ReactApplication::Host() noexcept {
   return m_host;
 }
 
-hstring ReactApplication::MainComponentName() noexcept {
-  return InstanceSettings().MainComponentName();
-}
-
-void ReactApplication::MainComponentName(hstring const &value) noexcept {
-  InstanceSettings().MainComponentName(value);
```
Comment (vmoroz) [inline vnext/Microsoft.ReactNative/ReactApplication.cpp:93]: >MainComponentName [](start = 21, length = 17)

Please also change the ReactInstanceSettings to remove this property.

**Comment by vmoroz** [comment]
>   Frame rootFrame{nullptr};

It seems that this code repeats what is already in the App template.

---
Refers to: vnext/Microsoft.ReactNative/ReactApplication.cpp:184 in 10e75fb. [](commit_id = 10e75fba5d6b0a6d51a90dc84f71347007191cf9, deletion_comment = False)

File: vnext/local-cli/generator-windows/templates/cpp/src/App.cpp (hunk 10 above)
Comment (asklar) [reply vnext/local-cli/generator-windows/templates/cpp/src/App.cpp:95]: yeah this is taken straight from the UWP [C++/WinRT] App code. 

Keeping this as is to stay in lockstep with that code and to simplify merges when we have to update it.

File: packages/microsoft-reactnative-sampleapps/windows/SampleAppCPP/packages.config (hunk 6 above)
Comment (asklar) [reply packages/microsoft-reactnative-sampleapps/windows/SampleAppCPP/packages.config:6]: @vmoroz  that's because you haven't merged that change yet ;-)

File: packages/microsoft-reactnative-sampleapps/windows/SampleAppCPP/SampleAppCpp.vcxproj (hunk 7 above)
Comment (asklar) [reply packages/microsoft-reactnative-sampleapps/windows/SampleAppCPP/SampleAppCpp.vcxproj:4]: This is where VS put it...

File: vnext/local-cli/generator-windows/templates/cpp/src/App.cpp (hunk 11 above)
Comment (asklar) [reply vnext/local-cli/generator-windows/templates/cpp/src/App.cpp:51]: not sure, but we could bring it up with the cppwinrt alias if you'd like to find out :)

File: vnext/local-cli/generator-windows/templates/cpp/src/MainPage.cpp [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+
+using namespace winrt;
+using namespace Windows::UI::Xaml;
+
+namespace winrt::<%=ns%>::implementation
+{
+    MainPage::MainPage()
+    {
+        InitializeComponent();
+        const auto& app = Application::Current().as<App>();
```
Comment (vmoroz) [inline vnext/local-cli/generator-windows/templates/cpp/src/MainPage.cpp:19]: >const auto& [](start = 8, length = 11)

Ideally we must not use references for temporaries.
Please change it to 'auto'.

File: vnext/local-cli/generator-windows/templates/cpp/src/App.cpp [hunk 14]
```diff
@@ -38,3 +42,39 @@ App::App() noexcept
...
+{
+    using super = AppT<App>;
```
Comment (vmoroz) [inline vnext/local-cli/generator-windows/templates/cpp/src/App.cpp:53]: >using super [](start = 4, length = 11)

It would be better to define in .h file in case if developers want to implement more interface.

File: vnext/local-cli/generator-windows/templates/cpp/src/App.cpp [hunk 15]
```diff
@@ -38,3 +42,39 @@ App::App() noexcept
...
+
+/// <summary>
+/// Invoked when Navigation to a certain page fails
+/// </summary>
+/// <param name="sender">The Frame which failed navigation</param>
+/// <param name="e">Details about the navigation failure</param>
+void App::OnNavigationFailed(IInspectable const&, NavigationFailedEventArgs const& e)
+{
+    throw hresult_error(E_FAIL, hstring(L"Failed to load Page ") + e.SourcePageType().Name);
+}
```
Comment (vmoroz) [inline vnext/local-cli/generator-windows/templates/cpp/src/App.cpp:80]: Could we avoid having these two methods in the default ReactNative template?
We already have a redbox and other facilities in RN for the lifetime management. I wonder if they bring any value.

File: vnext/local-cli/generator-windows/templates/cs/src/App.xaml.cs [hunk 16]
```diff
@@ -31,5 +32,23 @@ public App()
 
             InitializeComponent();
         }
+
+        /// <summary>
+        /// Invoked when the application is launched normally by the end user.  Other entry points
+        /// will be used such as when the application is launched to open a specific file.
+        /// </summary>
+        /// <param name="e">Details about the launch request and process.</param>
+        protected override void OnLaunched(LaunchActivatedEventArgs e)
```
Comment (vmoroz) [inline vnext/local-cli/generator-windows/templates/cs/src/App.xaml.cs:41]: >OnLaunched [](start = 32, length = 10)

Could we make the C++ and C# code match each other?
Here we check if frame is empty, but we do not do it in the C++ case.

File: vnext/local-cli/generator-windows/templates/cpp/src/MainPage.h [hunk 17]
```diff
@@ -0,0 +1,21 @@
+﻿#pragma once
+#include "MainPage.g.h"
+#include <winrt/Microsoft.ReactNative.h>
+
+// clang-format off
+namespace winrt::<%=ns%>::implementation
```
Comment (vmoroz) [inline vnext/local-cli/generator-windows/templates/cpp/src/MainPage.h:6]: ><%=ns%> [](start = 17, length = 7)

I wonder how it works: namespace here must have '::' separator, while in xaml file the '.' separator.
Do we support complex namespaces?

File: vnext/local-cli/generator-windows/templates/cpp/src/App.cpp (hunk 15 above)
Comment (asklar) [reply vnext/local-cli/generator-windows/templates/cpp/src/App.cpp:80]: @vmoroz the redbox will show up when something in the RN side goes wrong. Page navigation failures can happen even without any RN content / before RN content loads,  e.g. if the page hits an exception loading XAML markup, etc. so I think there is still value in having these here, but let me know if you disagree

File: vnext/local-cli/generator-windows/templates/cpp/src/MainPage.h (hunk 17 above)
Comment (asklar) [reply vnext/local-cli/generator-windows/templates/cpp/src/MainPage.h:6]: we set this to be either the name of the project, or what's passed as the `--ns` option:
```js
  const ns = options.ns || newProjectName;
```

File: packages/E2ETest/run_wdio.js [hunk 18]
```diff
@@ -117,6 +117,7 @@ function parseLogs() {
 }
 
 function printFailedTests(ft) {
+  console.log('Failed test cases: ');
```
Comment (NickGerleman) [inline packages/E2ETest/run_wdio.js:120]: Consider console.error and using Chalk to set a color.

**Comment by NickGerleman** [comment]
Marking as a breaking change since the template changes, and it seems like public APIs do as well (didn't look super closely tbh).
